dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/NettyFramework.java

```diff
@@ -0,0 +1,298 @@
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+import com.ibm.ws.ffdc.annotation.FFDCIgnore;
+import com.ibm.ws.kernel.feature.ServerStarted;
+import com.ibm.wsspi.kernel.service.utils.ServerQuiesceListener;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.group.ChannelGroup;
+import io.netty.channel.group.DefaultChannelGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.util.concurrent.GlobalEventExecutor;
+import io.openliberty.netty.internal.http.HttpServerInitializer;
+
+/**
+ * Sample (for now) implementation bundle for configuring and setting up Netty services
+ */
+@Component(configurationPid = "io.openliberty.netty.internal",
+    immediate = true, 
+    service = { NettyFramework.class, ServerQuiesceListener.class }, 
+    property = { "service.vendor=IBM" })
+public class NettyFramework implements ServerQuiesceListener {
+
+    private ChannelGroup serverChannels = null;
+
+    /** Reference to the executor service -- required */
+    private ExecutorService executorService = null;
+
+    /** server started logic borrowed from CHFWBundle */
+    private static AtomicBoolean serverCompletelyStarted = new AtomicBoolean(false);
+    private static Queue<Callable<?>> serverStartedTasks = new LinkedBlockingQueue<>();
+    private static Object syncStarted = new Object() {
+    }; // use brackets/inner class to make lock appear in dumps using class name
+
+    /**
+     * The TCP based bootstrap.
+     */
+    private ServerBootstrap serverBootstrap;
+
+    /**
+     * UDP based bootstrap.
+     */
+    private Bootstrap bootstrap;
+
+    private EventLoopGroup parentGroup;
+    private EventLoopGroup workerGroup;
+    private EventLoopGroup udpGroup;
+
+    @Activate
+    protected void activate(Map<String, Object> properties) {
+        serverChannels = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+//      udpGroup = new NioEventLoopGroup(0, executorService);
+
+        /*
+         * As a demo, this starts up a HTTP server at localhost:9081 which will return 
+         * an 200 OK response to any request
+         */
+        // use the executor service provided by Liberty
+        parentGroup = new NioEventLoopGroup(0, executorService);
```

Comment:
setExecutorService() will get called before this due to OSGi startup ordering?

Reply:
Right - the ExecutorService has a `MANDATORY` cardinality, so this bundle can't activate until that service is registered